diff --git a/source/75-twemoji.conf b/source/75-twemoji.conf
index 9ef221f..afcbb8a 100644
--- a/source/75-twemoji.conf
+++ b/source/75-twemoji.conf
@@ -101,21 +101,6 @@
 
     This obviously conflicts with other emoji fonts if you have them installed.
     -->
-    <match target="pattern">
-        <test qual="any" name="family"><string>Apple Color Emoji</string></test>
-        <edit name="family" mode="assign" binding="same"><string>Twemoji</string></edit>
-    </match>
-
-    <match target="pattern">
-        <test qual="any" name="family"><string>Segoe UI Emoji</string></test>
-        <edit name="family" mode="assign" binding="same"><string>Twemoji</string></edit>
-    </match>
-
-    <match target="pattern">
-        <test qual="any" name="family"><string>Segoe UI Symbol</string></test>
-        <edit name="family" mode="assign" binding="same"><string>Twemoji</string></edit>
-    </match>
-
     <match target="pattern">
         <test qual="any" name="family"><string>Noto Color Emoji</string></test>
         <edit name="family" mode="assign" binding="same"><string>Twemoji</string></edit>
