diff --git a/Makefile.in b/Makefile.in
index a39596f..3a14f38 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -33,8 +33,6 @@ prefix=	@prefix@
 srcdir=	@srcdir@
 sysconfdir= @sysconfdir@
 localdbdir= @localdbdir@
-user= @user@
-group= @group@
 
 CFLAGS=	@CFLAGS@ @DEFS@ -DSYSCONFDIR=\"${sysconfdir}\" \
 	-DLOCALDBDIR=\"${localdbdir}\"
@@ -103,13 +101,13 @@ $(srcdir)/ianaopts.h: gentab.pl bootp-dhcp-parameters
 
 install::
 	-mkdir -p $(sbindir) $(mandir)/man5 $(mandir)/man8
-	$(INSTALL_PROGRAM) -s -o $(user) -g $(group) $(TARGET) $(sbindir)
-	$(INSTALL_DATA) -o $(user) -g $(group) dhcp6c.8 $(mandir)/man8
-	$(INSTALL_DATA) -o $(user) -g $(group) dhcp6s.8 $(mandir)/man8
-	$(INSTALL_DATA) -o $(user) -g $(group) dhcp6relay.8 $(mandir)/man8
-	$(INSTALL_DATA) -o $(user) -g $(group) dhcp6ctl.8 $(mandir)/man8
-	$(INSTALL_DATA) -o $(user) -g $(group) dhcp6c.conf.5 $(mandir)/man5
-	$(INSTALL_DATA) -o $(user) -g $(group) dhcp6s.conf.5 $(mandir)/man5
+	$(INSTALL_PROGRAM) -s $(TARGET) $(sbindir)
+	$(INSTALL_DATA) dhcp6c.8 $(mandir)/man8
+	$(INSTALL_DATA) dhcp6s.8 $(mandir)/man8
+	$(INSTALL_DATA) dhcp6relay.8 $(mandir)/man8
+	$(INSTALL_DATA) dhcp6ctl.8 $(mandir)/man8
+	$(INSTALL_DATA) dhcp6c.conf.5 $(mandir)/man5
+	$(INSTALL_DATA) dhcp6s.conf.5 $(mandir)/man5
 
 includes::
 
diff --git a/configure.in b/configure.in
index 59dd84e..752f646 100644
--- a/configure.in
+++ b/configure.in
@@ -218,22 +218,6 @@ if test "$ts_cv_sa_len" = yes; then
 	AC_DEFINE([HAVE_SA_LEN], 1, [Define to 1 if you have a length field in struct sockaddr])
 fi
 
-dnl configure the ownership of the programs
-AC_MSG_CHECKING(the owner of the programs)
-AC_ARG_WITH(user,
-[  --with-user=USER	set the owner of the programs],
-    user="$withval", user="bin")
-AC_MSG_RESULT($user)
-AC_SUBST(user)
-
-dnl configure the group ownership of the programs
-AC_MSG_CHECKING(the group owner of the programs)
-AC_ARG_WITH(group,
-[  --with-group=GROUP	set the group owner of the programs],
-    group="$withval", group="bin")
-AC_MSG_RESULT($group)
-AC_SUBST(group)
-
 AC_CHECK_HEADERS(stdarg.h)
 
 AC_OUTPUT(Makefile)
