diff --git a/application/pages/instance/VersionPage.cpp b/application/pages/instance/VersionPage.cpp
index 54d75d0..93822f2 100644
--- a/application/pages/instance/VersionPage.cpp
+++ b/application/pages/instance/VersionPage.cpp
@@ -203,9 +203,8 @@ void VersionPage::updateVersionControls()
 {
     // FIXME: this is a dirty hack
     auto minecraftVersion = Version(m_profile->getComponentVersion("net.minecraft"));
-    bool newCraft = controlsEnabled && (minecraftVersion >= Version("1.14"));
     bool oldCraft = controlsEnabled && (minecraftVersion <= Version("1.12.2"));
-    ui->actionInstall_Fabric->setEnabled(newCraft);
+    ui->actionInstall_Fabric->setEnabled(true);
     ui->actionInstall_Forge->setEnabled(true);
     ui->actionInstall_LiteLoader->setEnabled(oldCraft);
     ui->actionReload->setEnabled(true);
