diff --git a/modules/gui/qt/components/playlist/vlc_model.cpp b/modules/gui/qt/components/playlist/vlc_model.cpp
index fcd5c88..14e24a1 100644
--- a/modules/gui/qt/components/playlist/vlc_model.cpp
+++ b/modules/gui/qt/components/playlist/vlc_model.cpp
@@ -89,7 +89,9 @@ QPixmap VLCModel::getArtPixmap( const QModelIndex & index, const QSize & size )
         {
             QImageReader reader( artUrl );
             reader.setDecideFormatFromContent( true );
-            artPix = QPixmap::fromImageReader( &reader ).scaled( size );
+            artPix = QPixmap::fromImageReader( &reader ).scaled( size,
+                                                        Qt::IgnoreAspectRatio,
+                                                        Qt::SmoothTransformation );
             if ( artPix.isNull() == false )
             {
                 QPixmapCache::insert( key, artPix );
